server/conn.go

```diff
@@ -579,26 +585,27 @@ func (cc *clientConn) Run() {
 		}
 
 		startTime := time.Now()
-		if err = cc.dispatch(data); err != nil {
+		if err = cc.dispatch(logutil.WithRecvTs(ctx, startTime.UnixNano()/int64(time.Millisecond)), data); err != nil {
 			if terror.ErrorEqual(err, io.EOF) {
 				cc.addMetrics(data[0], startTime, nil)
 				return
 			} else if terror.ErrResultUndetermined.Equal(err) {
-				log.Errorf("con:%d result undetermined error, close this connection %s",
-					cc.connectionID, errors.ErrorStack(err))
+				logutil.Logger(ctx).Error("result undetermined, close this connection", zap.Error(err))
 				return
 			} else if terror.ErrCritical.Equal(err) {
-				log.Errorf("con:%d critical error, stop the server listener %s",
-					cc.connectionID, errors.ErrorStack(err))
+				logutil.Logger(ctx).Error("critical error, stop the server listener", zap.Error(err))
 				metrics.CriticalErrorCounter.Add(1)
 				select {
 				case cc.server.stopListenerCh <- struct{}{}:
 				default:
 				}
 				return
 			}
-			log.Warnf("con:%d dispatch error:\n%s\n%q\n%s",
-				cc.connectionID, cc, queryStrForLog(string(data[1:])), errStrForLog(err))
+			logutil.Logger(ctx).Warn("dispatch error",
+				zap.String("connInfo", cc.String()),
+				zap.String("sql", queryStrForLog(string(data[1:]))),
+				zap.String("err", errStrForLog(err)),
```

Comment:
zap.Error(err) ?

Reply:
maybe better keep `errStrForLog` , it seems `errStrForLog` in purpose to suspend stack info for `duplicate key error`  https://github.com/pingcap/tidb/blob/2fe78619fc22bb96bceff19b220de9bd4e24d724/server/conn.go#L642